langchain/chat_models/vertexai.py

```diff
@@ -0,0 +1,141 @@
+"""Wrapper around Google VertexAI chat-based models."""
+from dataclasses import dataclass, field
+from typing import Any, Dict, List, Optional, Union
+
+from pydantic import root_validator
+
+from langchain.callbacks.manager import (
+    AsyncCallbackManagerForLLMRun,
+    CallbackManagerForLLMRun,
+)
+from langchain.chat_models.base import BaseChatModel
+from langchain.llms.vertexai import _VertexAICommon
+from langchain.schema import (
+    AIMessage,
+    BaseMessage,
+    ChatGeneration,
+    ChatResult,
+    HumanMessage,
+    SystemMessage,
+)
+from langchain.utilities.vertexai import raise_vertex_import_error
+
+
+@dataclass
+class _MessagePair:
+    """InputOutputTextPair represents a pair of input and output texts."""
+
+    question: HumanMessage
+    answer: AIMessage
+
+
+@dataclass
+class _ChatHistory:
+    """InputOutputTextPair represents a pair of input and output texts."""
+
+    history: List[_MessagePair] = field(default_factory=list)
+    system_message: Optional[SystemMessage] = None
+
+
+def _parse_chat_history(history: List[BaseMessage]) -> _ChatHistory:
+    """Parses a sequence of messages into history.
+
+    A sequency should be either (SystemMessage, HumanMessage, AIMessage,
+    HumanMessage, AIMessage, ...) or (HumanMessage, AIMessage, HumanMessage,
+    AIMessage, ...).
+    """
+    if not history:
+        return _ChatHistory()
+    first_message = history[0]
+    system_message = first_message if isinstance(first_message, SystemMessage) else None
+    chat_history = _ChatHistory(system_message=system_message)
+    messages_left = history[1:] if system_message else history
+    if len(messages_left) % 2 != 0:
+        raise ValueError(
+            f"Amount of messages in history should be even, got {len(messages_left)}!"
+        )
+    for question, answer in zip(messages_left[::2], messages_left[1::2]):
+        if not isinstance(question, HumanMessage) or not isinstance(answer, AIMessage):
+            raise ValueError(
+                "A human message should follow a bot one, "
+                f"got {question.type}, {answer.type}."
+            )
+        chat_history.history.append(_MessagePair(question=question, answer=answer))
+    return chat_history
+
+
+class ChatVertexAI(_VertexAICommon, BaseChatModel):
+    """Wrapper around Vertex AI large language models.
+
+    To use, you should have the
+    ``google.cloud.aiplatform.private_preview.language_models`` python package
+    installed.
+    """
+
+    model_name: str = "chat-bison"
+    chat: Any = None  #: :meta private:
```

Comment:
Is it possble to use the real type here?

Reply:
I don't think so since we should be able using langchain (and running unit tests) without installing any custom dependencies (like Vertex SDK).